pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java

```diff
@@ -127,7 +127,7 @@ public int filterEntriesForConsumer(Optional<MessageMetadata[]> optMetadataArray
         List<PositionImpl> entriesToRedeliver = CollectionUtils.isNotEmpty(entryFilters) ? new ArrayList<>() : null;
         for (int i = 0, entriesSize = entries.size(); i < entriesSize; i++) {
             Entry entry = entries.get(i);
-            if (entry == null) {
+            if (entry == null || entry.getDataBuffer() == null) {
```

Comment:
First, it's better to check `metadataAndPayload == null` to reduce one more invocation of `getDataBuffer()`.

Then, we should not resolve this issue by just adding a null check. `entry.getDataBuffer() == null` represents this entry has already been recycled. However, the corresponding item of a recycled entry in the entry list (`entries`) is usually set null, for example:

https://github.com/apache/pulsar/blob/926834ef8b2b57d3964aa7e9773e6245bcee861c/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java#L142-L143

Instead of null check simply, we should figure out in which case could `entries.get(i)` return a non-null entry that is recycled. In addition, I found your Pulsar version is 2.8.1.2, you might check the 2.8.1.2 source code first because there were many changes in master since that.

Reply:
OK ,thanks!